---
title: |
    Suggested TMT processing pipeline for PNNL DMS *phospho* data.

vignette: >
  %\VignetteIndexEntry{tmt_pipeline_for_PNNL_DMS_phos}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

output:
  BiocStyle::html_document:
     toc: true
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE)
# knitr::opts_chunk$set(echo=T, message=F, warning=F, fig.align='center', out.width='10cm')
```


`PlexedPiper` is the main library for the pipeline. `MSnID` package is a 
flexible way of handling MS/MS identifications. It handles MS/MS data
filtering part.

```{r libraries}
library(PlexedPiper)
library(PNNL.DMS.utils)
library(MSnID)
library(tidyverse)
```


# Data source
This vignette demonstrates processing usign PNNL's DMS as a data source.
PNNL DMS is based on MS SQL Server. Typically data is organized into data 
packages according to experiments. 
Data package that contains phospho data mirroring 
`PlexedPiperTestData` is number `3626`. Smaller version of the data 
package, suggested for testing purposes is `3625`.

```{r data package}
data_package_num <- 3625
```


```{r test_connection}
# if there is no connection to PNNL DMS, vignette is not compiled
if(!is_PNNL_DMS_connection_successful()){
  message("There is no connection to PNNL DMS. This code in this vignette won't be evaluated.")
  knitr::opts_chunk$set(eval = FALSE)
}
```


# Prepare MS/MS IDs

## Read the MS-GF+ output

First step is to determine MS-GF+ jobs using the data package number.
This simply reads parsed to text output of MS-GF+ search engine. The text files
are collated together and the resulting `data.frame` used to create MSnID object.

```{r msms}
msnid <- read_msgf_data_from_DMS(data_package_num)
show(msnid)
```


Phospho dataset involve Ascore jobs for improving phospho-site localization.
There should be one Ascore job per data package. The fetched object is a 
`data.frame` that links datasets, scans and original PTM localization to
newly suggested locations. Importantly it contains `AScore` column that 
signifies the confidence of PTM assingment. `AScore > 17` is considered 
confident.
```{r ascore}
ascore <- get_AScore_results(data_package_num)
msnid <- best_PTM_location_by_ascore(msnid, ascore)
```

Remove non-phospho. Need to be sure that phospho symbol is *!
```{r remove_glob}
msnid <- apply_filter(msnid, "grepl(\"\\\\*\", peptide)")
```


FDR filter at peptide level.
```{r FDR}
msnid <- filter_msgf_data_peptide_level(msnid, 0.01)
```

Remapping IDs from RefSeq to gene symbols both `MSnID` object and
the FASTA file use for the search.
```{r remap}
conv_tbl <- fetch_conversion_table(organism_name = "Rattus norvegicus", 
                                  from = "REFSEQ", to = "SYMBOL")

msnid <- remap_accessions_refseq_to_gene(msnid,
                                         conversion_table = conv_tbl)

# Remove missing accessions
psms(msnid) <- dplyr::filter(psms(msnid), !is.na(accession))

fst_dms_pth <- path_to_FASTA_used_by_DMS(data_package_num)
fst_dms_pth_2 <- remap_accessions_refseq_to_gene_fasta(
  fst_dms_pth, conversion_table = conv_tbl
)
```

Double checking FASTA file. Note, in case of multiple RefSeq mapping to the
same gene, we picked the RefSeq entry with the longest sequence. Unfortunately,
unlike UniProt, RefSeq does not report, which sequence should be 
considered as canonical. So to resolve ambiguity we resort to protein length.
```{r fasta_gene}
library(Biostrings)
fst <- readAAStringSet(fst_dms_pth_2)
```

Mapping sites to protein sequence. In a lot of cases it is easier to 
handle and interpret for example, `Hrc-T379t`, rather then peptide 
sequence `D.DDDNDGGST*ENVHQAHR.H` itself. 
```{r map_sites}
msnid <- map_mod_sites(msnid, fst, 
                       accession_col = "accession", 
                       peptide_mod_col = "peptide", 
                       mod_char = "*",
                       site_delimiter = "lower")
head(psms(msnid))
```

## Inference of parsimonius set

Inference of parsimonius step was skipped in this pipeline.
Remapping RefSeq IDs to gene symbols helped the situation of the
ambiguous assignment. Demo of inference should be available in other 
vignettes.


# Prepare MASIC reporter ion intensities

Fetching and preparing reporter intensities based on MASIC ouput.
```{r masic}
masic_data <- read_masic_data_from_DMS(data_package_num, 
                                       interference_score = T)
masic_data <- filter_masic_data(masic_data, 0.5, 0)
```


# Fetch study design tables

```{r study_design}
study_design_tables <- get_study_design_by_dataset_package(data_package_num)
fractions <- study_design_tables$fractions
samples <- study_design_tables$samples
references <- study_design_tables$references
```


# Create cross-tab

```{r cross_tab}
msnid <- apply_filter(msnid, "!isDecoy")
aggregation_level <- c("SiteID")
quant_cross_tab <- create_crosstab(msnid, 
                                   masic_data, 
                                   aggregation_level, 
                                   fractions, samples, references)
dim(quant_cross_tab)
head(quant_cross_tab)
```





